conda/cli/main_config.py

```diff
@@ -809,3 +833,92 @@ def execute_config(args, parser):
         from .common import stdout_json_success
 
         stdout_json_success(rc_path=rc_path, warnings=json_warnings, get=json_get)
+
+
+def migrate_pkgs(context, config: dict, json_warnings: list[str]):
+    """Migrate the pkgs/ directory from the root prefix to the user data directory.
+
+    Parameters
+    ----------
+    context : Context
+        Current execution context
+    config : dict
+        Configuration dict read from `.condarc`
+    json_warnings : list[str]
+        Warnings to be emitted
+    """
+    from ..base.context import user_data_pkgs
+
+    migrate_envs_or_pkgs(
+        context,
+        config,
+        "pkgs_dirs",
+        user_data_pkgs(),
+        context._legacy_root_prefix_pkgs(),
+        json_warnings,
+    )
+
+
+def migrate_envs(context, config: dict, json_warnings: list[str]):
+    """Migrate the envs/ directory from the root prefix to the user data directory.
+
+    Parameters
+    ----------
+    context : Context
+        Current execution context
+    config : dict
+        Configuration dict read from `.condarc`
+    json_warnings : list[str]
+        Warnings to be emitted
+    """
+    from ..base.context import user_data_envs
+
+    migrate_envs_or_pkgs(
+        context,
+        config,
+        "envs_dirs",
+        user_data_envs(),
+        context._legacy_root_prefix_envs(),
+        json_warnings,
+    )
+
+
+def migrate_envs_or_pkgs(
+    context,
+    config: dict,
+    key: str,
+    target_dir: os.PathLike,
+    legacy_dir: os.PathLike,
+    json_warnings: list[str],
+):
+    """Migrate either the envs/ or pkgs/ directories to the user data directory.
+
+    Parameters
+    ----------
+    context : Context
+        Current execution context
+    config : dict
+        Configuration dict read from `.condarc`
+    key : str
+    target_dir : os.PathLike
+        Target directory where files should be moved
+    legacy_dir : os.PathLike
+        Legacy directory whose contents need to be moved
+    json_warnings : list[str]
+        Warnings to be emitted
+    """
+    from .. import CondaError
+    from ..common.path.directories import copy_dir_contents
+
+    if key in config:
+        raise CondaError(
+            f"The conda configuration for {key} has explicitly set and"
+            "cannot be migrated."
+        )
+
+    copy_dir_contents(legacy_dir, target_dir)
```

Comment:
- For `pkgs/` we should try to hardlink first. Fallback to copy if not possible (e.g. different filesystem on target location).
- For `envs/`, we need to use `conda rename` (or `clone`? can't remember) because the paths will be different and the hardcoded locations (e.g. shebangs) need to be rewritten.

Reply:
For `pkgs/` I've made it attempt to hardlink first before resorting to copying, with the option to bypass hardlinking if `always_copy` is set. For `envs/`, `conda rename` seems to do the trick :+1: